### PR TITLE
Updated add_dataset_transform() to have flexible insertion point

### PR DESCRIPTION
**Pull request by growlix**
`utils.data.add_dataset_transform()` by default appends transformations to `dataset.tranform.transforms`. This is how algorithms such as `randaugment` and `augmix` insert themselves. Some datasets already have `to_tensor()` in the list of transforms (e.g. Cifar10), however `randaugment` and `augmix` need to be added *before* `to_tensor()`. This PR updates `add_dataset_transform()` to have an optional `location` argument that allows insertion before `to_tensor()`.

### CODE REVIEW

**Comment by hanlint** [comment]
Nice! Do we care to build a more generalized system? e.g.

```
before=ToTensor(), after=None
```

and allow someone to say.. insert this before `Normalize`?

File: composer/utils/data.py
Comment (hanlint) [inline composer/utils/data.py:105]:
```suggestion
        location (str): Where to insert the transform in the sequence of transforms. "end"
        will append to the end, "before_totensor" will insert before ToTensor(). Default: "end".
```

**Review by hanlint — COMMENTED** [review]
LGTM, but defer to @coryMosaicML for a review.

File: composer/utils/data.py
Comment (ravi-mosaicml) [inline composer/utils/data.py:113]: Can you do an if check and raise a valueerror instead of using assert? Asserts should not be used for data validation

**Comment by ravi-mosaicml** [comment]
> Nice! Do we care to build a more generalized system? e.g.
> 
> ```
> before=ToTensor(), after=None
> ```
> 
> and allow someone to say.. insert this before `Normalize`?

Let's leave this for now. But since this parameter would only have two values, could we make it a boolean - i.e. `is_tensor_transformation: bool = True`?

**Comment by growlix** [comment]
Made generalized as per @hanlint's request and idempotent as per @ravi-mosaicml's request

File: composer/utils/data.py
Comment (hanlint) [inline composer/utils/data.py:112]:
```suggestion
            transform string passed in 'location'. Ignored if location = "start" or "end". Default = "pre".
```

File: composer/utils/data.py
Comment (hanlint) [inline composer/utils/data.py:110]: Would it be safer to error out if `location` is not in the list?

File: composer/utils/data.py
Comment (hanlint) [inline composer/utils/data.py:132]: This is becoming sufficiently complex that perhaps we want to add some simple testing around this.

**Review by hanlint — APPROVED** [review]
LGTM, although we probably want to add a few tests.

File: composer/utils/data.py
Comment (growlix) [reply composer/utils/data.py:110]: Not every dataset has the same list of transforms, nor do they all have `ToTensor()` in the list of transforms. An error would require the user to provide `location` as a hyperparameter for the algorithm, which in turn would require them to know which transforms are in their dataset. This seems like a bit much. Perhaps a `log.warning()` displaying the list of transforms before the function returns?

Comment (hanlint) [reply composer/utils/data.py:108]: i'm ok with a warning, would prefer an error to avoid silent fail (e.g. was supposed to run before Normalize, but since no Normalize, it's now silently running at the end).

File: composer/utils/data.py
Comment (ravi-mosaicml) [inline composer/utils/data.py:139]: Curious when would this be true?

File: composer/utils/data.py
Comment (ravi-mosaicml) [inline composer/utils/data.py:97]: type annotation for dataset, transform, location

File: composer/utils/data.py
Comment (ravi-mosaicml) [reply composer/utils/data.py:108]: I find this API a bit confusing. Thoughts about replace the `location` and `pre_post` with two lists of transforms?

1. `pre_transformtions: Transformations`: Any transformations that, if present on `dataloader.transform`, will come before this new transformation
2. `post_transformations: Transformations`: Any transformations that, if present on `dataloader.transform`, will come this new transformation

If this is not possible (i.e. a post_transform already appears before a pre_transform), then a ValueError would be raised.

Comment (ravi-mosaicml) [reply composer/utils/data.py:108]: Alternatively, I think we may be over engineering this. Thoughts about a singular `is_tensor_transform: bool` argument? If it is True, then it should be inserted after ToTensor; if it is False, then it should be inserted before ToTensor. If there is no `ToTensor` present, then it's inserted at the end.

Do our algorithms depend on a specific ordering (other than for the tensors)?

File: composer/utils/data.py
Comment (ravi-mosaicml) [reply composer/utils/data.py:130]: Also, can this line be broken into multiple lines?

File: composer/utils/data.py
Comment (hanlint) [reply composer/utils/data.py:108]: I'm also OK with `before=ToTensor()` and/or `after=end` but maybe that makes the input validation a bit difficult. @growlix you have the design decision on the API, so make a decision based on these inputs and we'll go with it!

File: composer/utils/data.py
Comment (growlix) [reply composer/utils/data.py:154]: Not sure if you mean the `else` (L136) or the `if` (L139). L136 and below execute when `location` is not `"start"` or `"end"` (expected usage is that `location` is a `torchvision.transform`). L139 is true when the item in the list of transforms is of the same type as `location`, e.g. when they're both `transforms.ToTensor`.

**Comment by growlix** [comment]
I thought it'd be nice to have a flexible function that can insert a transform at location relative to other transforms, but you might be right about over-engineering. For now all we need is to insert before or after `ToTensor()` and to insert the transform somewhere else (default to end, or even always to end) if `ToTensor()` is not present.

File: composer/utils/data.py
Comment (growlix) [inline composer/utils/data.py:143]: How do I deal with all the 
```
Cannot access member "transform" for type "Dataset[Unknown]"
Member "transform" is unknown
```
errors from the static analysis?

Comment (hanlint) [reply composer/utils/data.py:143]: Hmm guess the static typechecker doesn't understand the guard you have upfront.

I'm not a typing expert, but I think you would use `Protocol` to implement a new type that has the attribute `transform`, and use that instead of `torch.utils.Dataset`, e.g. see [PEP 544](https://www.python.org/dev/peps/pep-0544/). 

But maybe defer to @ravi-mosaicml on this one.

File: composer/utils/data.py
Comment (hanlint) [inline composer/utils/data.py:127]: do we need the `textwrap.dedent` here and below when the errors are relatively short?

Comment (ravi-mosaicml) [reply composer/utils/data.py:127]: Nope, textwrap.dedent is only for multiline errors.

File: composer/utils/data.py
Comment (ravi-mosaicml) [reply composer/utils/data.py:160]: I think I missed that `t` could be a string or transform.

File: composer/utils/data.py
Comment (ravi-mosaicml) [reply composer/utils/data.py:143]: Can you add an `assert isinstance(dataset, VisionDataset)`?

File: composer/utils/data.py
Comment (ravi-mosaicml) [reply composer/utils/data.py:127]: Can we use check to ensure the dataset is a subclass of `torchvision.datasets.VisionDataset` (https://pytorch.org/vision/main/generated/torchvision.datasets.VisionDataset.html)?

e.g.

```python3
if not isinstance(dataset, torchvision.datasets.VisionDataset):
    raise TyepError(...)
```

File: composer/utils/data.py
Comment (ravi-mosaicml) [inline composer/utils/data.py:148]: For readabiity, can you remove the helper function and split it into two lines? Makes it more readable what's happening.

```suggestion
        log.warn(LOG_MESSAGE)
        added_transform = True
```
Same thing in the other places...

File: composer/utils/data.py
Comment (hanlint) [reply composer/utils/data.py:127]: Since we want to support custom datasets that have the `tranforms` attribute but not be properly inheriting from `VisionDataset`, we may not want to prematurely narrow the checks here.

Comment (ravi-mosaicml) [reply composer/utils/data.py:127]: Any custom dataset can inherit from `VisionDataset`. This function was designed around the structure of `VisionDataset`, so I think it should be limited to datasets of that type.

Comment (growlix) [reply composer/utils/data.py:127]: I don't think so, because Composer's synthetic datasets are of type `torch.utils.data.Dataset`, which would make them incompatible with this. But I think it could be reasonable to ensure that the dataset is a subclass of `torch.utils.data.Dataset`, which iirc is the most generic `dataset` class.

File: composer/utils/data.py
Comment (growlix) [reply composer/utils/data.py:144]: To set a flag for whether the transform was actually added, but probably best not to combine those functions.

File: composer/utils/data.py
Comment (ravi-mosaicml) [reply composer/utils/data.py:127]: Correct, our synthetic datasets do not support torchvision transforms, since the yield tensors directly. See https://github.com/mosaicml/composer/issues/346

**Comment by growlix** [comment]
I simplified `add_dataset_transform()` such that the only placement-related parameter is `is_tensor_transform`, and the new transform is placed before/after `ToTensor()` if `ToTensor()` is present in `dataset.transform`, and at the end if not. I also added a `SyntheticPILDataset` class that's nearly identical to the `SyntheticBatchPairDataset`, except that it generates/returns PIL images instead of tensors.

Also tests.

File: composer/utils/data.py
Comment (growlix) [reply composer/utils/data.py]: No worries!

File: composer/core/types.py
Comment (ravi-mosaicml) [inline composer/core/types.py:169]: Curious is this class being used? If not let's remove it?

File: composer/datasets/synthetic.py
Comment (ravi-mosaicml) [inline composer/datasets/synthetic.py:158]: Can you add a call to `super.__init__(root="", transform=transform)`

File: composer/utils/data.py
Comment (ravi-mosaicml) [inline composer/utils/data.py:109]: Docstring 
```suggestion
        dataset (VisionDataset): A torchvision-like dataset
```

File: composer/utils/data.py
Comment (ravi-mosaicml) [inline composer/utils/data.py:113]: For sphinx to parse it properly, lines 113-117 need to be indented.

File: composer/utils/data.py
Comment (ravi-mosaicml) [inline composer/utils/data.py:112]: RST uses double ticks for code literal blocks (I made this mistake wayy to many times, then the docs come out in italics :D) Not going to mark all, but please double check.

```suggestion
        is_tensor_transform (bool): Whether ``transform`` acts on data of the type
```

File: composer/utils/data.py
Comment (ravi-mosaicml) [inline composer/utils/data.py:113]: For auto-highlighting of methods, use these sphinx tags like `:meth:`

The leading `~` says to only show the last part (i.e. `ToTensor`) in the docs rather than the whole name (i.e. `torchvision.transforms.ToTensor`)

```suggestion
        torch.Tensor. If ``True``, and :meth:`~torchvision.transforms.ToTensor` is present in
```

Not going to comment on all, but let's enclose these in double-ticks so they will be formatted as code. Ideally we should reference it using the `:meth:` notation but can address that in a separate PR

File: composer/utils/data.py
Comment (ravi-mosaicml) [inline composer/utils/data.py:123]: Since the type annotation is a `VisionDataset`, this hasattr check is no longer needed.

File: composer/utils/data.py
Comment (ravi-mosaicml) [inline composer/utils/data.py:143]: Let's keep the comment, it's helpful :) 

Also, should this be an else?

```suggestion
    else: # transform is some other basic transform, join using Compose
```

File: composer/utils/data.py
Comment (ravi-mosaicml) [inline composer/utils/data.py:150]: I don't think this else would ever trigger?

File: docs/source/callbacks.rst
Comment (ravi-mosaicml) [inline docs/source/callbacks.rst:12]: Whoops, thanks!

File: composer/datasets/synthetic.py
Comment (ravi-mosaicml) [inline composer/datasets/synthetic.py:133]: Thanks for writing this! This pretty implements #346 :)

Jotting this down for another issue/PR, unless if you have the bandwidth to implement it tmrw :):

A lot of this logic seems very similar to the `SyntheticBatchPairDataset`.

Could the `SyntheticPILDataset` internally use a `SyntheticBatchPairDataset` to generate the tensors, and then wrap it in a PIL image (and apply the transformations) on `__getitem__`.? I think that would clean up the code.

**Review by ravi-mosaicml — APPROVED** [review]
Overall LGTM! See comments.

Can you try a docs build and make sure the docstrings format as expected? I tried to comment on some of the rst syntax errors but probably missed some. I don't want this to hold up the PR, though, since we're going to do a heavy docs scrub next week.

File: composer/datasets/synthetic.py
Comment (growlix) [reply composer/datasets/synthetic.py:133]: Oh wow, nice to have inadvertently resolved an outstanding issue :)
I'm not quite sure how to internally use a `SyntheticBatchPairDataset` to generate the data for `SyntheticPILDataset` because `SyntheticBatchPairDataset` generates the tensors upon the first call to `__getitem__`, and `SyntheticPILDataset` would need to override `__getitem__`. It might be possible to change `SyntheticBatchPairDataset` to generate the data on init, but the comment for `__getitem__` in `SyntheticBatchPairDataset` makes it seem as though the data need to be generated upon the first call to `__getitem__` so that they're on the correct gpu.

**Pull request merged**
